File: csharp/meetup/Meetup.cs
using System;
using System.Linq;

public enum Schedule
{
    Teenth,
    First,
    Second,
    Third,
    Fourth,
    Last
}

public class Meetup
{
    private readonly DateTime[] _days;

    public Meetup(int month, int year)
    {
        _days = Enumerable
            .Range(1, DateTime.DaysInMonth(year, month))
            .Select(day => new DateTime(year, month, day))
            .ToArray();
    }

    public DateTime Day(DayOfWeek dayOfWeek, Schedule schedule)
    {
        var candidates = _days.Where(dt => dt.DayOfWeek == dayOfWeek);

        return schedule switch
        {
            Schedule.First => candidates.First(),
            Schedule.Second => candidates.Skip(1).First(),
            Schedule.Third => candidates.Skip(2).First(),
            Schedule.Fourth => candidates.Skip(3).First(),
            Schedule.Last => candidates.Last(),
            Schedule.Teenth => candidates.First(dt => dt.Day >= 13 && dt.Day <= 19),
            _ => throw new ArgumentException("Invalid schedule")
        };
    }
}

File: kotlin/meetup/src/main/kotlin/Meetup.kt
import MeetupSchedule.*
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.Month
import java.time.Year

private val TEENTHS = (13..19)

class Meetup(month: Int, year: Int) {
    private val month = Month.of(month)
    private val year = Year.of(year)
    private val days = (1..this.month.length(this.year.isLeap))
    private val dates by lazy { days.map { LocalDate.of(year, month, it) } }

    fun day(dayOfWeek: DayOfWeek, schedule: MeetupSchedule): LocalDate {
        val candidates = dates.filter { it.dayOfWeek == dayOfWeek }

        return when (schedule) {
            FIRST, SECOND, THIRD, FOURTH -> candidates[schedule.ordinal]
            LAST -> candidates.last()
            TEENTH -> candidates.first { it.dayOfMonth in TEENTHS }
        }
    }
}
